### PR TITLE
loops-simple_programs_with_repetition-hard-q8

### PR DESCRIPTION
**Pull request by SamamthaH**
## PR Checklist
- [ ]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [ ]  Are the assertion fail messages good?
    - [ ]  Logical (does it make sense?)
    - [ ]  Readable (can a student understand what’s wrong?)
    - [ ]  Not too feed-y (does it give away the answer?)
    - [ ]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [ ]  Specific (does the fail message relate to the question?)
- [ ]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [ ]  EOF newlines
- [ ]  Do the comments for Parsons follow the correct convention?
- [ ]  Is the question being tested sufficiently?
    - [ ]  Are there representative cases to test the program?
    - [ ]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [ ]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”
---
## Website
- [ ] Is the question text on the Gamification document up to date? 
    > Make sure to include the sample output if there is one for the question

### CODE REVIEW

File: src/loops/simple_programs_with_repitition/hard/q8/VirtualSocialDistancing.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
+package loops.simple_programs_with_repitition.hard.q8;
+
+/*
+Write a program that will accept a string from the user that doesn't have spaces and then add spaces between the words
```
Comment (Hedgemon4) [inline src/loops/simple_programs_with_repitition/hard/q8/VirtualSocialDistancing.java:4]:
```suggestion
Write a program that will accept a string from the user that doesn't have any spaces and then add spaces between the words
```

File: src/loops/simple_programs_with_repitition/hard/q8/VirtualSocialDistancing.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+and store it in a separate string. Assume that the first letter of every word is a capital letter.
```
Comment (Hedgemon4) [inline src/loops/simple_programs_with_repitition/hard/q8/VirtualSocialDistancing.java:5]:
```suggestion
and stores it in a separate string. Assume that the first letter of every word is a capital letter.
```

File: src/loops/simple_programs_with_repitition/hard/q8/MainTest.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+public class MainTest extends BaseTest {
+    // Java
+    public Clause[] testSentence(){
+        TestOption.isInputTest = true;
+        TestOption.defaultInput = "HowCouldThisHappen";
+        return new Clause[]{
+                new StringLiteral("Enter a sentence:"),
+                new Optional(new StringLiteral(" ")),
+                new NewLine(),
+                new StringLiteral("Socially Distanced String: "),
```
Comment (Hedgemon4) [inline src/loops/simple_programs_with_repitition/hard/q8/MainTest.java:27]: Add the Optional with the space here as well

File: src/loops/simple_programs_with_repitition/hard/q8/MainTest.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+    public void runMain(){VirtualSocialDistancing.main(new String[0]);}
+
+    static Stream<Arguments> inputProvider(){
+        return Stream.of(Arguments.of("ByTalos", "By Talos"), Arguments.of("SCREAM", "S C R E A M"), Arguments.of("nothing", "nothing"), Arguments.of("W3lpIDone", "W3lp I Done"));
+    }
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    void correctlySeparatedString(String in, String seperated) throws InvalidClauseException{
```
Comment (Hedgemon4) [inline src/loops/simple_programs_with_repitition/hard/q8/MainTest.java:40]:
```suggestion
    void correctlySeparatesString(String in, String seperated) throws InvalidClauseException{
```

File: src/loops/simple_programs_with_repitition/hard/q8/MainTest.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        TestOption.incorrectStructureErrorMessage = "Your program does not correctly separate a string between capital letters.";
+        runWithInput(in, new Clause[]{
+                new StringLiteral(seperated)
+        });
+    }
+
+    @Test
+    void grabbedCorrectInput() throws InvalidClauseException {
+        TestOption.incorrectStructureErrorMessage = "Make sure to only grab input without pre-existing spaces.";
```
Comment (Hedgemon4) [inline src/loops/simple_programs_with_repitition/hard/q8/MainTest.java:49]:
```suggestion
        TestOption.incorrectStructureErrorMessage = "Your program should only accept input without any spaces.";
```

File: src/loops/simple_programs_with_repitition/hard/q8/MainTest.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    public void runMain(){VirtualSocialDistancing.main(new String[0]);}
+
+    static Stream<Arguments> inputProvider(){
+        return Stream.of(Arguments.of("ByTalos", "By Talos"), Arguments.of("SCREAM", "S C R E A M"), Arguments.of("nothing", "nothing"), Arguments.of("W3lpIDone", "W3lp I Done"));
+    }
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    void correctlySeparatedString(String in, String seperated) throws InvalidClauseException{
+        TestOption.incorrectStructureErrorMessage = "Your program does not correctly separate a string between capital letters.";
```
Comment (Hedgemon4) [inline src/loops/simple_programs_with_repitition/hard/q8/MainTest.java:41]:
```suggestion
        TestOption.incorrectStructureErrorMessage = "Your program does not correctly separate a string by capital letters.";
```

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/loops/simple_programs_with_repitition/hard/q8/MainTest.java (hunk 3 above)
Comment (SamamthaH) [reply src/loops/simple_programs_with_repitition/hard/q8/MainTest.java:27]: I don't know, the space is visible in the sample and it's good programming format to leave a space there.

File: src/loops/simple_programs_with_repitition/hard/q8/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+                new NewLine(),
+                new StringLiteral("Socially Distanced String: "),
+                new PlaceHolder()
+        };
+    }
+
+    public void runMain(){VirtualSocialDistancing.main(new String[0]);}
+
+    static Stream<Arguments> inputProvider(){
+        return Stream.of(Arguments.of("ByTalos", "By Talos"), Arguments.of("SCREAM", "S C R E A M"), Arguments.of("nothing", "nothing"), Arguments.of("W3lpIDone", "W3lp I Done"));
```
Comment (Hedgemon4) [inline src/loops/simple_programs_with_repitition/hard/q8/MainTest.java:35]: I think you should also add another case with special characters (?, !, $, etc. )

File: src/loops/simple_programs_with_repitition/hard/q8/MainTest.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    @MethodSource("inputProvider")
+    void correctlySeparatesString(String in, String seperated) throws InvalidClauseException{
+        TestOption.incorrectStructureErrorMessage = "Your program does not correctly separate a string by capital letters.";
+        runWithInput(in, new Clause[]{
+                new StringLiteral(seperated)
+        });
+    }
+
+    @Test
+    void grabbedCorrectInput() throws InvalidClauseException {
```
Comment (Hedgemon4) [inline src/loops/simple_programs_with_repitition/hard/q8/MainTest.java:48]:
```suggestion
    void readsInputCorrectly() throws InvalidClauseException {
```

File: src/loops/simple_programs_with_repitition/hard/q8/VirtualSocialDistancing.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
+package loops.simple_programs_with_repitition.hard.q8;
+
+/*
+Write a program that will accept a string from the user that doesn't have any spaces and then add spaces between the words
```
Comment (Hedgemon4) [inline src/loops/simple_programs_with_repitition/hard/q8/VirtualSocialDistancing.java:4]:
```suggestion
Write a program that will accept a string from the user that doesn't have any spaces, adds spaces between the words, 
```

**Review by Hedgemon4 — APPROVED** [review]

**Pull request merged**